py/mkrules.mk

```diff
@@ -84,9 +102,20 @@ $(OBJ): | $(HEADER_BUILD)/qstrdefs.generated.h $(HEADER_BUILD)/mpversion.h $(OBJ
 # - else, if list of newer prerequisites ($?) is not empty, then process just these ($?)
 # - else, process all source files ($^) [this covers "make -B" which can set $? to empty]
 # See more information about this process in docs/develop/qstr.rst.
-$(HEADER_BUILD)/qstr.i.last: $(SRC_QSTR) $(QSTR_GLOBAL_DEPENDENCIES) | $(QSTR_GLOBAL_REQUIREMENTS)
+# Since C and C++ files are preprocessed separately in the same target make sure
+# both steps can always run by having always at least one source file.
+$(HEADER_BUILD)/noqstr.cpp: $(QSTR_GLOBAL_REQUIREMENTS)
+	$(ECHO) "GEN $@"
+	$(Q)$(TOUCH) $@
+
+$(HEADER_BUILD)/noqstr.c: $(QSTR_GLOBAL_REQUIREMENTS)
+	$(ECHO) "GEN $@"
+	$(Q)$(TOUCH) $@
+
+$(HEADER_BUILD)/qstr.i.last: $(HEADER_BUILD)/noqstr.cpp $(HEADER_BUILD)/noqstr.c $(SRC_QSTR) $(QSTR_GLOBAL_DEPENDENCIES) | $(QSTR_GLOBAL_REQUIREMENTS)
 	$(ECHO) "GEN $@"
-	$(Q)$(CPP) $(QSTR_GEN_CFLAGS) $(if $(filter $?,$(QSTR_GLOBAL_DEPENDENCIES)),$^,$(if $?,$?,$^)) >$(HEADER_BUILD)/qstr.i.last
+	$(Q)$(CPP) $(QSTR_GEN_CFLAGS) $(HEADER_BUILD)/noqstr.c $(filter-out %.cpp,$(if $(filter $?,$(QSTR_GLOBAL_DEPENDENCIES)),$^,$(if $?,$?,$^))) >$(HEADER_BUILD)/qstr.i.last
+	$(Q)$(CPP) $(QSTR_GEN_CXXFLAGS) $(HEADER_BUILD)/noqstr.cpp $(filter %.cpp,$(if $(filter $?,$(QSTR_GLOBAL_DEPENDENCIES)),$^,$(if $?,$?,$^))) >>$(HEADER_BUILD)/qstr.i.last
```

Comment:
If both .c and .cpp go through the same preprocessor, isn't it possible to run just one command for them?  Or is it because of the CXXFLAGS that that doesn't work?

Also, does having both lines here now require a `c++`/`g++` compiler to be installed even if the code is pure C?

Reply:
> isn't it possible to run just one command for them?

Had it like that first but that causes 2 problems:
- default CFLAGS usually has -std=gnu11 or so which leads to compiler warning that the flag isn't valid for C++ code
- when a user decides to use CXXFLAGS to enable code which contains QSTRs, preprocessing with CFLAGS wouldn't see that code and miss the QSTRs

> does having both lines here now require a c++/g++ compiler to be installed even if the code is pure C

Yes, but that's also the result of me not knowing how to express in make 'only run this line when there are effectively C++ files'.

Reply:
Both of those points could be perhaps addressed by filtering out the required items, eg:
```
$(CPP) $(filter-out -std=%,$(QSTR_GEN_CFLAGS)) $(filter -D%,$(QSTR_GEN_CXXFLAGS)) ...
```
but that seems pretty fragile... 

Two options I can see are:
1. Allow `makeqstrdefs.py` to take multiple `qstr.i.last` files (as you suggested).
2. Add a new sub-command to `makeqstrdefs.py` that does the call to cpp with the correct flags and source files (possibly multiple calls of there are both C and C++ code).  This would eliminate the complex `$(if $?,$?,$^)` stuff and also be useful for potential cmake builds that cannot specify this complex set of dependencies.

Reply:
Second option seems the way to go then (shell function could also work I guess) - but just to make sure we're on the same page here: since you mention getting rid of the complex stuff this would be a call like
```
(Q)$(PYTHON) $(PY_SRC)/makeqstrdefs.py pp $(CPP) $(HEADER_BUILD) $(HEADER_BUILD)/qstr.i.last "$?" "$^" "$(QSTR_GLOBAL_DEPENDENCIES)"
```
and makeqstrdefs.py does the 'if any global dependencies changed then PP all, else if there are newer dependencies PP only those, else PP all' logic?

Reply:
Yes, exactly like that.

But I also thought of another way, option 3: repeat this rule twice, once for C and C++, then add a new rule to cat the outputs together.  Schematically:
```
$(HEADER_BUILD)/qstr.i.last.c: ...
        # parse C files

$(HEADER_BUILD)/qstr.i.last.cpp: ...
        # parse C++ files

$(HEADER_BUILD)/qstr.i.last: $(HEADER_BUILD)/qstr.i.last.c  $(HEADER_BUILD)/qstr.i.last.cpp
        cat $^ > $@
```
Although that might again need special handling when there are no C++ files... and it won't help with eventual cmake support.  So if you have the time to do option 2 then feel free to attempt it.

Reply:
Yes I already tried that option 3 and it has the same problem. And concatenating the 2 files together, of which only 1 might just have been built meaning the other one is an older version is also not ideal.

The Python solution is fairly easy, and looks like it works in all possible situations, however the quoting of arguments proves to be rather non-trivial. If the Makefile has
```
CFLAGS = -DSTM32_HAL_H='<stm32$(MCU_SERIES)xx_hal.h>' -DMP_CONFIGFILE='"mpconfigport_nanbox.h"' -DFFCONF_H=\"lib/oofatfs/ffconf.h\"
```
which are actual flags being used across variuos ports, then passing this as `"$(CFLAGS)"` to Python the according call to `print(sys.argv[1])` shows
```
-DSTM32_HAL_H='<stm32xx_hal.h>' -DMP_CONFIGFILE='mpconfigport_nanbox.h' -DFFCONF_H="lib/oofatfs/ffconf.h"
```
I don't understand where the double quotes around mpconfigport_nanbox.h went, so I see no other option then to change the Makefile (and erase all the single quotes in Python to deal with '<stm32xx_hal.h>'), then turn that into a list of strings to pass to `subprocess.call` which also needs some tricks since it should split on spaces, but not when they are in quotes, and the quotes must be retained (which e.g. `shlext.split` cannot do). Currently everything except the Zephyr minimal port builds though.
Given the above problems I'm thinking there must be an easier way (perhaps having the Makefile generate code fragments which Python can execute?) or I'm missing something.

Reply:
It seems to work OK for me if the CFLAGS are not quoted, ie:
```
$(Q)$(PYTHON) $(PY_SRC)/makeqstrdefs.py pp $(CPP) cflags $(QSTR_GEN_CFLAGS) cxxflags $(QSTR_GEN_CXXFLAGS) source $(SRC_QSTR) dependencies $(QSTR_GLOBAL_DEP) ...
```
Since flags always start with `-` and source files always have an extension, it should be possible to separate out those parts based on the sentinels cflags/cxxflags/source/dependencies within the long list of command line args.

Reply:
Ok good catch, leaving the quotes passes everything as it should be. Went with the sentinel solution and that works well. 

Added a coverage test; was doubting whether to put the module in examples/ but left it in variants/coverage/ after all, any preference?

Reply:
> Added a coverage test; was doubting whether to put the module in examples/ but left it in variants/coverage/ after all, any preference?

I think it'll get lost in `ports/unix/variants/coverage/`.  If it's a pure coverage test then it could be `ports/unix/coveragecpp.cpp` and mirror how `ports/unix/coverage.c` works (or, actually, `coverage.c` could be converted to a user C module...).  Or if it's an example then better to go in `examples/usercmodule/cppexample/`.

Really there are two things to test here: 1) including C++ code in a normal build (eg `ports/unix/coveragecpp.cpp`); 2) making a user C++ module (eg `examples/usercmodule/cppexample/`, which could also sit next to a C user module like `examples/usercmodule/cexample/`).  We could have both, but if you want to limit the scope of this PR then simplest thing to do would be to move your existing example to `examples/`.

Reply:
Ok. More than once I found myself going to the C module docs to copy-paste the code there into respective files, so I think it would be good if a complete example is readily available (and mentioned in the docs). Indeed probably best to have separate C and C++ examples. I'll add a `ports/unix/coveragecpp.cpp` as well, good idea.

Reply:
@dpgeorge didn't think about the fact that coverag.c and coverage.cpp have the same name apart from the extension, so they cannot both be compiled into coverage.o nor use the same dependency file, etc. Going to take some time to fix that (and keeping ESP32 working, not sure how much it relies on having .o for it's C++ code)

Reply:
But can't the source files be `coverage.c` and `coveragecpp.cpp`, so they generate different `.o` files?  We'll just have the restriction that all C and C++ source files must have unique base names, but I think that's a very reasonable restriction.

Reply:
Sure, that's definitely a lot easier.

All examples added.